### PR TITLE
Regexp acceptor, fixes #200

### LINKED ISSUE #200: Regex Acceptor

### PR DESCRIPTION
**Pull request by Protryon**
Ported from shapesecurity/shift-parser-js#403

Supersedes #201.

### CODE REVIEW

File: src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java [hunk 1]
```diff
@@ -0,0 +1,728 @@
...
+public class PatternAcceptor {
+
+    public final String pattern;
+    public final boolean unicode;
+
+    private static final String[] decimalDigits = new String[]{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9"};
+    private static final String[] octalDigits = new String[]{"0", "1", "2", "3", "4", "5", "6", "7"};
+    private static final String[] hexDigits = new String[]{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "a", "b", "c", "d", "e", "f", "A", "B", "C", "D", "E", "F"};
+    private static final String syntaxCharacters = "^$\\.*+?()[]{}|";
+    private static final String[] syntaxCharacterArray = "^$\\.*+?()[]{}|".split("");
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java]: `syntaxCharacters.split("");`?

File: src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java [hunk 2]
```diff
@@ -0,0 +1,728 @@
...
+
+        public Context() {
+            this.index = 0;
+            this.backreferenceNames = ImmutableSet.emptyUsingEquality();
+            this.groupingNames = ImmutableSet.emptyUsingEquality();
+            this.backreferences = ImmutableSet.emptyUsingEquality();
+            this.capturingGroups = 0;
+        }
+
+        public boolean addGrouping() {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java]: This function is only used once. Let's inline it.

File: src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java [hunk 3]
```diff
@@ -0,0 +1,728 @@
...
+            this.capturingGroups++;
+            return true;
+        }
+
+        public void backreference(int num) {
+            this.backreferences = this.backreferences.put(num);
+        }
+
+        public boolean verifyBackreferences() {
+            if (unicode) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java]: This is an instance var, so let's use `this.` so we know it's not static or local.

**edit:** Let's just do that consistently everywhere.

File: src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java [hunk 4]
```diff
@@ -0,0 +1,728 @@
...
+
+    private static final HashMap<String, Integer> controlEscapeCharacterValues = constructControlEscapeCharacterValues();
+
+    private static final String[] controlEscapeCharacters = controlEscapeCharacterValues.keySet().toArray(new String[0]);
+    
+    private class Context {
+        private int index;
+        private ImmutableSet<String> backreferenceNames;
+        private ImmutableSet<String> groupingNames;
+        private ImmutableSet<Integer> backreferences;
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java]: We don't need to keep all of these in a set, we just need to keep the maximum value we've seen.

File: src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java [hunk 5]
```diff
@@ -0,0 +1,728 @@
...
+        public Context() {
+            this.index = 0;
+            this.backreferenceNames = ImmutableSet.emptyUsingEquality();
+            this.groupingNames = ImmutableSet.emptyUsingEquality();
+            this.backreferences = ImmutableSet.emptyUsingEquality();
+            this.capturingGroups = 0;
+        }
+
+        public boolean addGrouping() {
+            this.capturingGroups++;
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java]: I prefer to always use prefix increment/decrement. It has simpler semantics.

File: src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java [hunk 6]
```diff
@@ -0,0 +1,728 @@
...
+                return Maybe.empty();
+            }
+            if (!unicode) {
+                return Maybe.of(new String(Character.toChars(pattern.charAt(this.index))));
+            }
+            return Maybe.of(new String(Character.toChars(pattern.codePointAt(this.index))));
+        }
+
+        public void skipCodePoint() {
+            this.index += this.nextCodePoint().orJust("").length();
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java]:
```java
this.nextCodePoint().forEach(cp -> { this.index += cp.length; });
```

File: src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java [hunk 7]
```diff
@@ -0,0 +1,723 @@
...
+
+    private static final String[] decimalDigits = new String[]{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9"};
+    private static final String[] octalDigits = new String[]{"0", "1", "2", "3", "4", "5", "6", "7"};
+    private static final String[] hexDigits = new String[]{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "a", "b", "c", "d", "e", "f", "A", "B", "C", "D", "E", "F"};
+    private static final String syntaxCharacters = "^$\\.*+?()[]{}|";
+    private static final String[] syntaxCharacterArray = syntaxCharacters.split("");
+    private static final String extendedSyntaxCharacters = "^$.*+?()[|";
+    private static final String[] controlCharacters = new String[]{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z", "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"};
+
+    private static HashMap<String, Integer> constructControlEscapeCharacterValues() {
```
Comment (bakkot) [inline src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java]: Do this with a `static` block:

```java
private static final HashMap<String, Integer> controlEscapeCharacterValues = new HashMap<>();
static {
  controlEscapeCharacterValues.put("f", (int) '\f');
  // etc
}
```

File: src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java [hunk 8]
```diff
@@ -0,0 +1,723 @@
...
+    private boolean acceptRegex() {
+        Context context = new Context();
+        if (!acceptDisjunction(context, Maybe.empty())) {
+            return false;
+        }
+        return context.verifyBackreferences();
+    }
+
+    @SafeVarargs
+    private final <B> F<Context, Maybe<B>> maybeLogicalOr(F<Context, Maybe<B>>... expressions) {
```
Comment (bakkot) [inline src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java]: Let's call this `anyOf`, as in the JS implementation.

File: src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java [hunk 9]
```diff
@@ -0,0 +1,721 @@
...
+
+        public String collect(@Nonnull String[]... stringArrays) {
+            return collect(Maybe.empty(), stringArrays);
+        }
+
+        public String collect(Maybe<Integer> limit, @Nonnull String[]... stringArrays) {
+            StringBuilder stringBuilder = new StringBuilder();
+
+            outer:
+            for (int i = 0; limit.isNothing() || limit.fromJust() > i; ++i) {
```
Comment (bakkot) [inline src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java]: Let's pull out the `isNothing` and `fromJust` from the test, so it doesn't evaluate every time.

Comment (Protryon) [reply src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java]: I see I am not the only one with no faith in the JVM. :D

File: src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java [hunk 10]
```diff
@@ -0,0 +1,690 @@
...
+            state -> acceptCharacterClassEscape(state) ? Maybe.of(Maybe.empty()) : Maybe.empty(),
+            state -> acceptCharacterEscape(state).map(Maybe::of)
+        ).apply(superState);
+    }
+
+    private Maybe<Maybe<Integer>> acceptClassAtomNoDash(State state) {
+        if (state.eat("\\")) {
+			return this.anyOf(
+					this::acceptClassEscape,
+                subState -> subState.backtrackOnFailureMaybe(subState2 -> {
```
Comment (bakkot) [inline src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java:607]: Whitespace here is weird.

File: src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java [hunk 11]
```diff
@@ -0,0 +1,690 @@
...
+    }
+
+    private Maybe<Maybe<Integer>> acceptClassAtom(State state) {
+        if (state.eat("-")) {
+            return Maybe.of(Maybe.of((int)"-".charAt(0)));
+        }
+        return acceptClassAtomNoDash(state);
+    }
+
+    private Maybe<Maybe<Integer>> finishClassRange(State state, Maybe<Integer> atom) {
```
Comment (bakkot) [inline src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java:630]: Can we have a comment here explaining what each of the three kinds of return value here mean?

Comment (bakkot) [reply src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java:630]: Ah, never mind, I see you have that below. But ideally move it up to above the first function with that return type.

File: src/test/java/com/shapesecurity/shift/es2017/Serialization/DeserializerTest.java [hunk 12]
```diff
@@ -136,7 +136,7 @@ public void testLiterals() throws IllegalAccessException, NoSuchMethodException,
         testHelperFromScriptCode("0");
         testHelperFromScriptCode("1.5");
         testHelperFromScriptCode("/[a-z]/i");
-        testHelperFromScriptCode("/(?!.){0,}?/u");
+        testHelperFromScriptCode("/(?!.){0,}?/");
```
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2017/Serialization/DeserializerTest.java:139]: Can we add a test with a valid regex with `u`?

File: src/test/java/com/shapesecurity/shift/es2017/parser/expressions/literals/LiteralRegExpExpressionTest.java [hunk 13]
```diff
@@ -1,18 +1,239 @@
...
+        "/t{5/",
+        "/{5,10X}/",
+        "/{5,X}/",
+        "/{5.}/",
+        "/{dfwfdf}/",
+        "/|.||.|/",
+        "/|/",
+        "/}*/",
+        "/[\\99-\\98]/",
+        "/[\\99-\\100]/",
```
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2017/parser/expressions/literals/LiteralRegExpExpressionTest.java:160]: Let's get the tests in [this commit](https://github.com/shapesecurity/shift-regexp-acceptor-js/commit/5f254de0eef1178de9f023b9ef9b10c530c1a7d8#diff-c1d9835a59f549d9ee3db0bfb144e125R187) over here as well.

**Review by bakkot — APPROVED** [review]
LGTM other than missing tests.

**Pull request merged**